Allow to reorder engine prioritization with `engines` key in _quarto.yml

jkrumbiegel opened the pull request:
## Description

This PR adds the ability to change the order in which quarto checks if a markdown file or a language is claimed by an engine. This is necessary to allow users of the native julia engine to use it project-wide, without having to specify `engine: julia` in each notebook's frontmatter. This is because for backwards compatibility, the native julia engine neither claimed jl script files nor julia language blocks in qmd files, so that jupyter would continue to claim them as usual.

Now, the julia engine does claim these properties, however, jupyter also does, and because the engines are checked in order, jupyter will always win by default. However, by specifying `engines: ['julia']` in `_quarto.yml`, the native julia engine is checked first and therefore wins.

This PR therefore fixes https://github.com/quarto-dev/quarto-cli/issues/10034 and https://github.com/quarto-dev/quarto-cli/issues/11305

## Todos

I'm marking this WIP as I haven't added any tests, yet. First, I'd need to know that this approach is valid, for example if the location of the key in the config is correct. If it is, maybe some quarto-internal schema needs to be extended with this key as well, I couldn't quite make sense of that code when I tried to add `engines` below the `project` key at first. The top-level doesn't seem to get validated so I implemented the feature there.

## Checklist

I have (if applicable):

- [x] filed a [contributor agreement](https://github.com/quarto-dev/quarto-cli/blob/main/CONTRIBUTING.md).
- [x] referenced the GitHub issue this PR closes
- [ ] updated the appropriate changelog in the PR
- [x] ensured the present test suite passes
- [x] added new tests
- [ ] created a separate documentation PR in [Quarto's website repo](https://github.com/quarto-dev/quarto-web/) and linked it to this PR

jkrumbiegel committed:
Allow to reorder engine prioritization with `engines` key in _quarto.yml

cscheid commented:
Hey, this is going to be an excellent addition to Quarto, thanks!

> I'm marking this WIP as I haven't added any tests, yet. 

Sounds good, but this _is_ something we'll definitely need tests for.

> First, I'd need to know that this approach is valid, for example if the location of the key in the config is correct. 

> If it is, maybe some quarto-internal schema needs to be extended with this key as well, I couldn't quite make sense of that code when I tried to add engines below the project key at first. 

Did you try adding it to [`src/resources/schema/project.yml`](https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/schema/project.yml). That should have worked. Specifically, I'd have added a top-level entry to the file like so:

````yaml
- name: engines
  schema:
    arrayOf: string
  description: "..."
````

jkrumbiegel committed:
add engines to schema

jkrumbiegel committed:
add tests for `engines` key effect

jkrumbiegel committed:
add test for invalid engine

jkrumbiegel committed:
use testing/asserts import path

jkrumbiegel commented:
@cscheid I have added tests. I also added an entry to the schema yml file you suggested, although I can't tell if it had an effect on anything. I was already allowed to put the `engines` key at top level before, so nothing I can observe changed.

jkrumbiegel commented:
Small bump :)

cderv commented on src/execute/engine.ts:
```diff
@@ -143,6 +147,38 @@ export function markdownExecutionEngine(markdown: string, flags?: RenderFlags) {
 }
 
+function reorderEngines(project: ProjectContext) {
+  // TODO: add `engines` to schema
```
This is done by [`8e76196` (#11807)](https://github.com/quarto-dev/quarto-cli/pull/11807/commits/8e761969f58a27d66e96758d73901dc10d1ca61e) right ?

jkrumbiegel replied:
ah then that's why it had no effect, I will remove the todo comment

cderv commented on src/execute/julia.ts:
```diff
@@ -109,7 +118,13 @@ export const juliaEngine: ExecutionEngine = {
 
   markdownForFile(file: string): Promise<MappedString> {
-    return Promise.resolve(mappedStringFromFile(file));
+    if (isJupyterPercentScript(file)) {
```
Should'nt we use `isJuliaPercentScript()` ? Probably equivalent, but would maybe be clearer for future us 🤔 

Or am I missing something ?

jkrumbiegel replied:
no I think you're right, this was copied over from the jupyter implementation and might as well use `isJuliaPercentScript`

jkrumbiegel committed:
remove stale todo

jkrumbiegel committed:
change `isJupyterPercentScript` to `isJuliaPercentScript`

mcanouil commented:
Am I mistaken to thing this will also fix?

- #3157

jkrumbiegel commented:
> Am I mistaken to think this will also fix?

It will not directly fix #3157, it does nothing to change the behavior of the `engine` key in `_quarto.yml`. But the idea was that this might not be the right fix anyway, because if you use multiple different languages in your project, it wouldn't make sense to set say `julia` as the engine for all of them. With the reordering, you'd ideally be able to prioritize all engines that matter to you if you use multiple languages.

Anything I can do to move this along @cscheid :) ?

jkrumbiegel commented:
One more bump :)

jkrumbiegel commented:
Would be great to get this one into 1.7 if that's still possible :) Please let me know if there's anything I can do to move this PR forward

cscheid commented:
Yes, totally possible. Sorry, my last month was a mess of travel and personal stuff. Let's definitely merge this ahead of 1.7.

cscheid commented:
@jkrumbiegel I'm going to merge this so your users can start testing. Do you think you could draft a docs PR at https://github.com/quarto-dev/quarto-web/?

cscheid reviewed (approved):
This looks good but we will definitely want documentation since it's not otherwise a very discoverable feature.
